src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs

```diff
@@ -0,0 +1,203 @@
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Collections.Sequences;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers
+{
+    public readonly partial struct ReadOnlyBuffer 
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        static bool TryGetBuffer(Position begin, Position end, out ReadOnlyMemory<byte> data, out Position next)
+        {
+            var segment = begin.Segment;
+
+            switch (segment)
+            {
+                case null:
+                    data = default;
+                    next = default;
+                    return false;
+
+                case IMemoryList<byte> bufferSegment:
+                    var startIndex = begin.Index;
+                    var endIndex = bufferSegment.Memory.Length;
+
+                    if (segment == end.Segment)
+                    {
+                        endIndex = end.Index;
+                        next = default;
+                    }
+                    else
+                    {
+                        var nextSegment = bufferSegment.Next;
+                        if (nextSegment == null)
+                        {
+                            if (end.Segment != null)
+                            {
+                                ThrowHelper.ThrowInvalidOperationException(ExceptionResource.EndCursorNotReached);
+                            }
+
+                            next = default;
+                        }
+                        else
+                        {
+                            next = new Position(nextSegment, 0);
+                        }
+                    }
+
+                    data = bufferSegment.Memory.Slice(startIndex, endIndex - startIndex);
+
+                    return true;
+
+
+                case OwnedMemory<byte> ownedMemory:
+                    data = ownedMemory.Memory.Slice(begin.Index, end.Index - begin.Index);
+
+                    if (segment != end.Segment)
+                    {
+                        ThrowHelper.ThrowInvalidOperationException(ExceptionResource.EndCursorNotReached);
+                    }
+
+                    next = default;
+                    return true;
+
+                case byte[] array:
+                    data = new Memory<byte>(array, begin.Index, end.Index - begin.Index);
+
+                    if (segment != end.Segment)
+                    {
+                        ThrowHelper.ThrowInvalidOperationException(ExceptionResource.EndCursorNotReached);
+                    }
+                    next = default;
+                    return true;
+            }
+
+            ThrowHelper.ThrowNotSupportedException();
+            next = default;
+            return false;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        static Position Seek(Position begin, Position end, long bytes, bool checkEndReachable = true)
```

Comment:
internal

Reply:
same: why not private?

Reply:
So the tests from tests/System.Buffers.Primitives.Tests/ReadableBufferFacts.cs  do not require copied code or being moved.